### ISSUE TITLE
Package with dash in its name fails to install with Yarn and NPM

### ISSUE THREAD

**Issue by raido** [verdaccio/verdaccio#1429]
**Describe the bug**

This package: https://yarnpkg.com/en/package/@ember-data/-build-infra is sub dependency of ember-data package. When trying to install ember-data@3.11 or @ember-data/-build-infra package directly served by Verdaccio, Verdaccio will return 404 error.

**To Reproduce**

Make sure package is not already on local cache.

`yarn add @ember-data/-build-infra` or `npm install @ember-data/-build-infra`

**Expected behavior**

Package to be successfully installed.

**Environment information**

Verdaccio 4.2.0
Yarn v1.17.3
NPM v6.4.1

**Screenshots**

Seems like there is no tarball stored on disk at all.

<img width="693" alt="Screenshot 2019-08-05 at 17 12 37" src="https://user-images.githubusercontent.com/659945/62471087-9131f280-b7a4-11e9-8a8e-20ac62afd9ca.png">

**Comment by raido** [comment]
Related: #1400

**Comment by bertdeblock** [comment]
Running into the same issue as well.

**Comment by juanpicado** [comment]
Did not we ported the patch to the last release 3.x already ?

**Comment by raido** [comment]
For the allowed names yes. How ever missing tarball issue is a new bug. It’s not returning 404 because of package name being invalid but rather actually failing to match/store/provide tarball itself.

**Comment by raido** [comment]
Current issue is for v4.2.0 as stated above.

**Comment by juanpicado** [comment]
Ahhh ! sorry, I did not saw it ... thanks :)

**Comment by raido** [comment]
Any updates on this?

**Comment by nickschot** [comment]
Seeing the same thing.

**Comment by eduardoweiland** [comment]
Seems to be caused by these lines in [validateName](https://github.com/verdaccio/verdaccio/blob/b1cd0704b925186f8f1b9c275572a996deb56b71/src/api/middleware.ts#L36) and [validatePackage](https://github.com/verdaccio/verdaccio/blob/b1cd0704b925186f8f1b9c275572a996deb56b71/src/api/middleware.ts#L49).

**Comment by schleumer** [comment]
I do have the same issue with 4.3.4, in my case is with package `regenerator-runtime`.

**Comment by raido** [comment]
As workaround I scope only internal package to point to Verdaccio in .npmrc and the rest is installed from public NPM registry.

**Comment by schleumer** [comment]
@raido, done that! Worked like a charm and maybe saved some server bandwidth hehe, i've reported just to bump the issue.

**Comment by kkincade** [comment]
I'm also seeing this issue specific to ember-data. This is a pretty significant issue for my company, and as a result we will need to find an alternative to Verdaccio if it is not fixed soon.

```
error Couldn't find package "@ember-data/-build-infra@3.14.0" required by "ember-data@~3.14.0" on the "npm" registry.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
Error: Couldn't find package "@ember-data/-build-infra@3.14.0" required by "@ember-data/serializer@3.14.0" on the "npm" registry.
```

**Comment by bertdeblock** [comment]
@kkincade Have you tried raido's suggestion => https://github.com/verdaccio/verdaccio/issues/1429#issuecomment-546588271?

**Comment by kkincade** [comment]
@bertdeblock I had to figure out how to do this workaround in Yarn. It does indeed seem to work after clearing out any previous registry settings that I had manually configured.

Here is what I used for my .yarnrc file for those interested.
```
"@my-company:registry" "http://verdaccio.our.internal.domain.com"
```

**Comment by raido** [comment]
Here is an example how to set it up with .npmrc 

https://docs.npmjs.com/files/npmrc.html#comments

**Comment by DanielRuf** [comment]
Current state in 4.4.0 / master:

> error An unexpected error occurred: "http://localhost:4873/@ember-data%2f-build-infra/-/-build-infra-3.14.1.tgz: Request failed \"404 Not Found\"

**Comment by DanielRuf** [comment]
The mentioned lines seem to be a working solution but then a test fails:

> Summary of all failing tests
> FAIL  test/unit/modules/api/api.spec.ts (14.58s)
>  ● endpoint unit test › Registry API Endpoints › should test ping api › should test endpoint /-/ping

**Comment by juanpicado** [comment]
Shipped on `v4.4.1` powered by @DanielRuf
